src/lightning_habana/utils/hpu_distributed.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) 2023 Habana Labs, Ltd. an Intel Company
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from lightning_utilities.core.imports import module_available
+from torch import Tensor
+
+if module_available("lightning"):
+    from lightning.fabric.utilities.distributed import _sync_ddp
+    from lightning.fabric.utilities.rank_zero import rank_zero_warn
+    from lightning.fabric.utilities.types import ReduceOp
+elif module_available("pytorch_lightning"):
+    from lightning_fabric.utilities.types import ReduceOp
+    from lightning_fabric.utilities.distributed import _sync_ddp
+
+from typing import Any, Optional, Union
+
+# Supported ReduceOps: https://docs.habana.ai/en/latest/API_Reference_Guides/HCCL_APIs/C_API.html#hcclredop-t
+supported_reduce_ops = {
+    "sum": ReduceOp.SUM,
+    "min": ReduceOp.MIN,
+    "max": ReduceOp.MAX,
+}
+
+
+def _distributed_available() -> bool:
+    return (
+        torch.distributed.is_available()
+        and torch.distributed.is_initialized()
+        and torch.distributed.get_backend() == "hccl"
+    )
+
+
+def check_reduce_op_support(reduce_op: Union[ReduceOp, str]) -> bool:
+    """Function to check if reduce_op is supported with hccl backend."""
+    reduce_op = reduce_op.lower() if isinstance(reduce_op, str) else reduce_op
+    if reduce_op in ("mean", "avg") or reduce_op == ReduceOp.AVG:
+        rank_zero_warn(f"{reduce_op} is not supported with HCCL. Going to simulate it")
+        return True
+    if reduce_op in supported_reduce_ops or any(reduce_op is op for op in supported_reduce_ops.values()):
+        return True
+    return False
+
+
+def _sync_ddp_if_available(
+    result: Tensor, group: Optional[Any] = None, reduce_op: Optional[Union[ReduceOp, str]] = "sum"
+) -> Tensor:
+    """Function to reduce a tensor across worker processes during distributed training.
+
+    Args:
+        result: The value to sync and reduce (typically tensor or number)
+        group: The process group to gather results from. Defaults to all processes (world)
+        reduce_op: The reduction operation. Defaults to sum.
+
+    Return:
+        reduced value
+
+    """
+    if _distributed_available():
+        assert check_reduce_op_support(
+            reduce_op
+        ), f"Unsupported ReduceOp {reduce_op}. Only 'sum', 'min', and 'max' are supported with HCCL"
```

Comment:
```suggestion
        ), f"Unsupported ReduceOp {reduce_op}. Supported ops in HCCL are: {', '.join(supported_reduce_ops)}"
```

Reply:
Perhaps the assertion should be made a proper exception. And it could also be moved into `check_reduce_op_support`. Otherwise I would call it `is_reduce_op_supported`. I also suggest to make it protected.